# Importing Libraries & Loading the Data

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split #Import train_test_split from sklearn.model_selection module 
import warnings
warnings.filterwarnings('ignore')

In [144]:
customer = pd.read_excel("Question 3 Data Set.xlsx")

# EDA & Preprocessing, Groupby Functions

In [92]:
customer.head()

Sales Zone Billing Date  Customer Code               Distributor  \
0       EAST   2019-08-02        2111797  UNIVERSAL TRADE & AGENCY   
1       EAST   2019-09-10        2111797  UNIVERSAL TRADE & AGENCY   
2       EAST   2019-08-12        2111797  UNIVERSAL TRADE & AGENCY   
3       EAST   2019-04-06        2111797  UNIVERSAL TRADE & AGENCY   
4       EAST   2019-05-11        2111797  UNIVERSAL TRADE & AGENCY   

    Material Code  Product Type Pack Classification Pack Type Pack Size  \
0  FBR08A30416B10  BASMATI RICE                  BP       BAG     25 KG   
1  FBR04A20102B10  BASMATI RICE                  BP       BAG     25 KG   
2  FBR08A30416B10  BASMATI RICE                  BP       BAG     25 KG   
3  FBR08A10416B10  BASMATI RICE                  BP       BAG     25 KG   
4  FBR08A30416B10  BASMATI RICE                  BP       BAG     25 KG   

   YTD Vol (MT)  
0        157.35  
1        128.00  
2         67.00  
3         64.00  
4         64.00

In [93]:
customer.tail()

Sales Zone Billing Date  Customer Code                  Distributor  \
91104       EAST   2019-12-27        2116123                   BIPLAB PAL   
91105       EAST   2019-06-03        2117167      RELIANCE RETAIL LTD CNC   
91106       EAST   2019-04-18        2119186  SHIV SHANKAR EDIBLE PVT LTD   
91107       EAST   2019-08-02        2111229          M/S BALAJEE TRADING   
91108       EAST   2020-03-23        2124345  AMAZON RETAIL INDIA PVT LTD   

           Material Code          Product Type Pack Classification Pack Type  \
91104  FATTA5091P07V1N2E           CHAKKI ATTA                  CP     POUCH   
91105  FATTA5091B09N2ES1           CHAKKI ATTA                  CP       BAG   
91106    FATTA5093B15N2E           CHAKKI ATTA                  BP       BAG   
91107       FNUGG3061P07  REGULAR SOYA NUGGETS                  CP     POUCH   
91108    FATTA5091B10N2E           CHAKKI ATTA                  CP       BAG   

      Pack Size  YTD Vol (MT)  
91104      1 KG         -3.75  
91105     10 KG         -3.99  
91106     50 KG         -5.00  
91107      1 KG         -5.46  
91108     10 KG        -12.00

In [94]:
customer.shape

(91109, 10)

In [95]:
customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 91109 entries, 0 to 91108
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Sales Zone           91109 non-null  object        
 1   Billing Date         91109 non-null  datetime64[ns]
 2   Customer Code        91109 non-null  int64         
 3   Distributor          91109 non-null  object        
 4   Material Code        91109 non-null  object        
 5   Product Type         91109 non-null  object        
 6   Pack Classification  91109 non-null  object        
 7   Pack Type            91109 non-null  object        
 8   Pack Size            91109 non-null  object        
 9   YTD Vol (MT)         91109 non-null  float64       
dtypes: datetime64[ns](1), float64(1), int64(1), object(7)
memory usage: 7.0+ MB


In [96]:
customer.columns

Index(['Sales Zone', 'Billing Date', 'Customer Code', 'Distributor',
       'Material Code', 'Product Type', 'Pack Classification', 'Pack Type',
       'Pack Size', 'YTD Vol (MT)'],
      dtype='object')

In [97]:
customer.describe()

Customer Code  YTD Vol (MT)
count   9.110900e+04  91109.000000
mean    2.115771e+06      0.613356
std     4.864312e+03      2.174308
min     2.110342e+06    -12.000000
25%     2.111849e+06      0.040000
50%     2.114810e+06      0.108000
75%     2.117300e+06      0.352000
max     2.128682e+06    157.350000

In [98]:
customer['Sales Zone'].unique()

array(['EAST'], dtype=object)

In [99]:
customer['Distributor'].unique()

array(['UNIVERSAL TRADE & AGENCY', 'AMAZON RETAIL INDIA PVT LTD',
       'SUBUDHI TRADERS', 'MANPASAND AGRO FOOD PVT LTD', 'SHANKAR STORE',
       'ARUNA ENTERPRISE', 'ROUNAK DISTRIBUTORS', 'SAMLESWARI TRADERS',
       'JAGDISH PRASAD ANAND KUMAR', 'MORE RETAIL PRIVATE LIMITED',
       'OM PRAKASH GOURI SHANKAR', "SPENCER'S RETAIL LTD", 'JAIN STORES',
       'R K ENTERPRISES', 'ANIL TRADERS', 'JAIN BHANDAR',
       'DEVEE DISTRIBUTORS', 'JAI KRISHNA AGRIBIZ PVT LTD',
       'SHAW TRADERS', 'MATHURA PRASAD MANOJ KUMAR',
       'SHREE MAHADEV TRADING COMPANY', 'GOYAL TRADING CO',
       'MAHABIR STORE', 'BAGWAN', 'SREE VENKATESWARA ENTERPRISES',
       'SHREE GANGA ENTERPRISES', 'AGRO POINT', 'BENGAL TRADING COMPANY',
       'BANSAL TRADING COMPANY', 'JAYASHREE ENTERPRISES',
       'MANJUSHREE ENTERPRISES', 'AMIT ENTERPRISE', 'MAA SARADA STORES',
       'UDAYNATH TRADING & CO', 'MAHALAXMI BHANDAR',
       'UNIVERSAL AGENCY HOUSE', 'SRIKANTA STORE', 'SALASAR FOODS',
       'MAHALAXMI TRAD

In [100]:
customer['Material Code'].unique()

array(['FBR08A30416B10', 'FBR04A20102B10', 'FBR08A10416B10',
       'FBR04A10102B10', 'FATTA5091B10N2E', 'FBR07HR0101B10',
       'FNUGG4062B11', 'FATTA5091B09N2E', 'FATTA5091P07V1N2E',
       'FNR02HR0422B10', 'FBR01TI0416B10', 'FBR01HR0608P04V1',
       'FPULS8133B15', 'FPULS7143B13', 'FNR01HR0435B10',
       'FATTA5093B12N2E', 'FATTA5094B09CUP', 'FBESA1024B10',
       'FPULS8143B15', 'FNR03HR0439B10', 'FBESA1013B10', 'FBESA2013B12',
       'FBESA1013B15', 'NSOAPC01W010101B', 'FBESA1013B14OR',
       'FBESA1013B14', 'FPULS8123B15', 'FPULS13123B15', 'FBR01HR0106B10',
       'FNR03HR0438B10', 'FNR04HR0141B08', 'FBR04DU0102B10',
       'FATTA5093B12CUP', 'FBR01HR0416B10', 'FNUGG3062P07',
       'FBR01HR0108B10', 'FBESA1011B10', 'FBR01HR0433B10', 'FBESA1014B10',
       'FATTA5091P09CUP', 'FATTA5093B15N2E', 'FBR07HR0101B08',
       'FBR01HR0101P04', 'FBR01HR0608B10', 'FATTA5091B09N2ES1',
       'FBR04TI0102B10', 'FBESA1023B10', 'FBESA1021P07',
       'FATTA5094B10CUP', 'FNUGG3061P07S1', '

In [101]:
customer['Product Type'].unique()

array(['BASMATI RICE', 'CHAKKI ATTA', 'NON-GMO SOYA NUGGETS',
       'NON BASMATI', 'MATAR', 'MASOOR', 'CHANA BESAN', 'MATAR BESAN',
       'SOAP CAKE', 'CHANA', 'REGULAR SOYA NUGGETS', 'RICE BRAN', 'BLEND',
       'MAIDA', 'SUJI', 'GROUND NUT', 'ARHAR', 'KHICHADI'], dtype=object)

In [102]:
customer['Pack Classification'].unique()

array(['BP', 'CP'], dtype=object)

In [103]:
customer['Pack Size'].unique()

array(['25 KG', '10 KG', '20 KG', '5 KG', '1 KG', '50 KG', '30 KG',
       '75 GM', '35 KG', '5 LTR', '1 LTR', '500 GM', '50 GM', '200 GM',
       '15 LTR', '90 GM', '2 LTR', '100 GM', '200 ML'], dtype=object)

In [104]:
customer['Pack Type'].unique()

array(['BAG', 'POUCH', 'PAPER WRAP', 'JAR', 'PET'], dtype=object)

In [105]:
customer['Material Code'].value_counts()

FATTA5091P07V1N2E    6351
FATTA5091B09N2E      6153
FNUGG3061P04S1       4883
FNUGG3061P01S1       4605
ERRBO01P03V1         4595
                     ... 
EBLEN75P03              1
FBR04HR0101B10          1
FBR02A10102B08          1
ERRBO01J08S1            1
FBR01MO0102B08V1        1
Name: Material Code, Length: 201, dtype: int64

In [106]:
customer['Customer Code'].value_counts()

2117491    1352
2122193    1159
2116824     746
2117013     718
2124650     707
           ... 
2123908       1
2124629       1
2112672       1
2118702       1
2125083       1
Name: Customer Code, Length: 894, dtype: int64

In [107]:
# Apply min, max, std, mean functions on all numeric columns
min_values = customer.min()
max_values = customer.max()
std_values = customer.std()
mean_values = customer.mean()

# Print the results
print("Minimum values:\n", min_values)
print("\nMaximum values:\n", max_values)
print("\nStandard deviation values:\n", std_values)
print("\nMean values:\n", mean_values)

Minimum values:
 Sales Zone                            EAST
Billing Date           2019-04-03 00:00:00
Customer Code                      2110342
Distributor                       8 ELEVEN
Material Code                   EBLEN03J03
Product Type                         ARHAR
Pack Classification                     BP
Pack Type                              BAG
Pack Size                             1 KG
YTD Vol (MT)                         -12.0
dtype: object

Maximum values:
 Sales Zone                            EAST
Billing Date           2020-03-31 00:00:00
Customer Code                      2128682
Distributor                         ZENITH
Material Code             NSOAPC01W010201B
Product Type                          SUJI
Pack Classification                     CP
Pack Type                            POUCH
Pack Size                            90 GM
YTD Vol (MT)                        157.35
dtype: object

Standard deviation values:
 Billing Date     104 days 07:35:53.723198482
Cus

# Conversion

In [147]:
customer = pd.get_dummies(customer, columns=['Sales Zone'], prefix='Sales_Zone')

In [148]:
customer.head(3)

Billing Date  Customer Code               Distributor   Material Code  \
0   2019-08-02        2111797  UNIVERSAL TRADE & AGENCY  FBR08A30416B10   
1   2019-09-10        2111797  UNIVERSAL TRADE & AGENCY  FBR04A20102B10   
2   2019-08-12        2111797  UNIVERSAL TRADE & AGENCY  FBR08A30416B10   

   Product Type Pack Classification Pack Type Pack Size  YTD Vol (MT)  \
0  BASMATI RICE                  BP       BAG     25 KG        157.35   
1  BASMATI RICE                  BP       BAG     25 KG        128.00   
2  BASMATI RICE                  BP       BAG     25 KG         67.00   

   Sales_Zone_EAST  
0                1  
1                1  
2                1

In [149]:
print(customer['YTD Vol (MT)'].dtype)

float64


In [150]:
print(customer['Sales_Zone_EAST'].dtype)

uint8


In [151]:
print(customer['Pack Size'].dtype)

object


In [152]:
customer = pd.get_dummies(customer, columns=['Pack Classification'], prefix='Pack Classification')

In [153]:
customer.head(3)

Billing Date  Customer Code               Distributor   Material Code  \
0   2019-08-02        2111797  UNIVERSAL TRADE & AGENCY  FBR08A30416B10   
1   2019-09-10        2111797  UNIVERSAL TRADE & AGENCY  FBR04A20102B10   
2   2019-08-12        2111797  UNIVERSAL TRADE & AGENCY  FBR08A30416B10   

   Product Type Pack Type Pack Size  YTD Vol (MT)  Sales_Zone_EAST  \
0  BASMATI RICE       BAG     25 KG        157.35                1   
1  BASMATI RICE       BAG     25 KG        128.00                1   
2  BASMATI RICE       BAG     25 KG         67.00                1   

   Pack Classification_BP  Pack Classification_CP  
0                       1                       0  
1                       1                       0  
2                       1                       0

In [154]:
conversion_factors = {
    'KG': 1000,
    'GM': 1,
    'LTR': 1000,
    'ML': 1
}


In [155]:
def convert_to_grams(pack_size):
    global conversion_factors
    size, unit = pack_size.split()
    size = float(size)
    factor = conversion_factors[unit]
    return size * factor


In [156]:
customer['Pack Size (g)'] = customer['Pack Size'].apply(convert_to_grams)

In [157]:
customer.head(3)

Billing Date  Customer Code               Distributor   Material Code  \
0   2019-08-02        2111797  UNIVERSAL TRADE & AGENCY  FBR08A30416B10   
1   2019-09-10        2111797  UNIVERSAL TRADE & AGENCY  FBR04A20102B10   
2   2019-08-12        2111797  UNIVERSAL TRADE & AGENCY  FBR08A30416B10   

   Product Type Pack Type Pack Size  YTD Vol (MT)  Sales_Zone_EAST  \
0  BASMATI RICE       BAG     25 KG        157.35                1   
1  BASMATI RICE       BAG     25 KG        128.00                1   
2  BASMATI RICE       BAG     25 KG         67.00                1   

   Pack Classification_BP  Pack Classification_CP  Pack Size (g)  
0                       1                       0        25000.0  
1                       1                       0        25000.0  
2                       1                       0        25000.0

In [158]:
customer['Pack Size (g)'].dtype

dtype('float64')

## Merge

In [159]:
customer['Total Pack'] = customer['Pack Size (g)'] * customer['YTD Vol (MT)']

In [160]:
customer.head(3)

Billing Date  Customer Code               Distributor   Material Code  \
0   2019-08-02        2111797  UNIVERSAL TRADE & AGENCY  FBR08A30416B10   
1   2019-09-10        2111797  UNIVERSAL TRADE & AGENCY  FBR04A20102B10   
2   2019-08-12        2111797  UNIVERSAL TRADE & AGENCY  FBR08A30416B10   

   Product Type Pack Type Pack Size  YTD Vol (MT)  Sales_Zone_EAST  \
0  BASMATI RICE       BAG     25 KG        157.35                1   
1  BASMATI RICE       BAG     25 KG        128.00                1   
2  BASMATI RICE       BAG     25 KG         67.00                1   

   Pack Classification_BP  Pack Classification_CP  Pack Size (g)  Total Pack  
0                       1                       0        25000.0   3933750.0  
1                       1                       0        25000.0   3200000.0  
2                       1                       0        25000.0   1675000.0

## Pivot Table

In [161]:
pivot_table = pd.pivot_table(customer, values='Total Pack', index='Customer Code', aggfunc='sum')

print(pivot_table)

                 Total Pack
Customer Code              
2110342        110777.90500
2110343        942620.00000
2110345         91035.01805
2110347        198837.47200
2110348         52255.05800
...                     ...
2128514           111.45000
2128542         38855.13980
2128558          2730.99600
2128680          1187.71100
2128682          2933.45500

[894 rows x 1 columns]


## Melt

In [162]:
melted = pd.melt(customer, id_vars=['Customer Code'], var_name='Pack Size', value_name='Pack Size (g)')

print(melted)

         Customer Code     Pack Size        Pack Size (g)
0              2111797  Billing Date  2019-08-02 00:00:00
1              2111797  Billing Date  2019-09-10 00:00:00
2              2111797  Billing Date  2019-08-12 00:00:00
3              2111797  Billing Date  2019-04-06 00:00:00
4              2111797  Billing Date  2019-05-11 00:00:00
...                ...           ...                  ...
1093303        2116123    Total Pack              -3750.0
1093304        2117167    Total Pack             -39900.0
1093305        2119186    Total Pack            -250000.0
1093306        2111229    Total Pack              -5460.0
1093307        2124345    Total Pack            -120000.0

[1093308 rows x 3 columns]


In [163]:
customer.head(2)

Billing Date  Customer Code               Distributor   Material Code  \
0   2019-08-02        2111797  UNIVERSAL TRADE & AGENCY  FBR08A30416B10   
1   2019-09-10        2111797  UNIVERSAL TRADE & AGENCY  FBR04A20102B10   

   Product Type Pack Type Pack Size  YTD Vol (MT)  Sales_Zone_EAST  \
0  BASMATI RICE       BAG     25 KG        157.35                1   
1  BASMATI RICE       BAG     25 KG        128.00                1   

   Pack Classification_BP  Pack Classification_CP  Pack Size (g)  Total Pack  
0                       1                       0        25000.0   3933750.0  
1                       1                       0        25000.0   3200000.0

In [164]:
# Apply min, max, std, mean functions on all numeric columns
min_values = customer['Total Pack'].min()
max_values = customer['Total Pack'].max()
std_values = customer['Total Pack'].std()
mean_values = customer['Total Pack'].mean()

# Print the results
print("Minimum values:\n", min_values)
print("\nMaximum values:\n", max_values)
print("\nStandard deviation values:\n", std_values)
print("\nMean values:\n", mean_values)

Minimum values:
 -250000.0

Maximum values:
 3933750.0

Standard deviation values:
 48154.89327603854

Mean values:
 5524.599449301543


# ML Technique

In [125]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import OrdinalEncoder


# Perform ordinal encoding on 'Customer Code'
ordinal_encoder = OrdinalEncoder()
customer['Customer Code Encoded'] = ordinal_encoder.fit_transform(customer[['Customer Code']])

# Discretize 'Total Pack' into classes for classification
customer['Total Pack Class'] = pd.cut(customer['Total Pack'], bins=3, labels=['Low', 'Medium', 'High'])

# Split data into independent (X) and dependent (y) variables
X = customer[['Customer Code Encoded']]
y = customer['Total Pack Class']

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize and train a Logistic Regression model
model = LogisticRegression()
model.fit(X_train, y_train)

# Make predictions
y_pred = model.predict(X_test)

# Evaluate model performance
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)

print(f"Accuracy: {accuracy}")
print("Classification Report:")
print(report)


Accuracy: 0.9998902425639337
Classification Report:
              precision    recall  f1-score   support

         Low       1.00      1.00      1.00     18220
      Medium       0.00      0.00      0.00         2

    accuracy                           1.00     18222
   macro avg       0.50      0.50      0.50     18222
weighted avg       1.00      1.00      1.00     18222



## with Oversampling 

In [166]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import StandardScaler, OrdinalEncoder
from imblearn.over_sampling import SMOTE


# Perform ordinal encoding on 'Customer Code'
ordinal_encoder = OrdinalEncoder()
customer['Customer Code Encoded'] = ordinal_encoder.fit_transform(customer[['Customer Code']])

# Scale 'Total Pack' for better handling in logistic regression
scaler = StandardScaler()
customer['Total Pack Scaled'] = scaler.fit_transform(customer[['Total Pack']])

# Define bins based on quantiles
bins = pd.qcut(customer['Total Pack Scaled'], q=3, labels=['Low', 'Medium', 'High'], duplicates='drop')
customer['Total Pack Class'] = bins

# Check class distribution
print("Class Distribution:")
print(customer['Total Pack Class'].value_counts())

# Split data into independent (X) and dependent (y) variables
X = customer[['Customer Code Encoded', 'Total Pack Scaled']]
y = customer['Total Pack Class']

# Apply SMOTE to handle class imbalance
smote = SMOTE(random_state=42)
X_resampled, y_resampled = smote.fit_resample(X, y)

# Split resampled data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_resampled, y_resampled, test_size=0.2, random_state=42)

# Initialize and train a Logistic Regression model
model = LogisticRegression(random_state=42)
model.fit(X_train, y_train)

# Make predictions
y_pred = model.predict(X_test)

# Evaluate model performance
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)

print(f"Accuracy: {accuracy}")
print("Classification Report:")
print(report)


Class Distribution:
Low       31576
High      30282
Medium    29251
Name: Total Pack Class, dtype: int64
Accuracy: 0.6232450121397657
Classification Report:
              precision    recall  f1-score   support

        High       1.00      0.82      0.90      6261
         Low       0.48      0.55      0.51      6302
      Medium       0.48      0.50      0.49      6383

    accuracy                           0.62     18946
   macro avg       0.65      0.62      0.64     18946
weighted avg       0.65      0.62      0.63     18946



## Analysis

Balanced Class Distribution: The dataset shows a relatively balanced distribution across the 'Total Pack Class' categories, with approximately 31,576 samples for 'Low', 30,282 samples for 'High', and 29,251 samples for 'Medium'. This balanced distribution ensures that the model has been trained on a diverse set of examples from each class, contributing to a more robust and generalized model.

Reasonable Accuracy: The model achieves an accuracy of 0.623, which indicates that it correctly predicts the 'Total Pack Class' for approximately 62.3% of the samples in the test dataset. This level of accuracy suggests that the model is performing above random chance and is making meaningful predictions based on the input features.

Precision and Recall Scores: The classification report reveals positive precision and recall scores across all classes. Specifically, for the 'High' class, the precision is perfect at 1.00, indicating that all predictions for this class were correct. The recall for 'High' is 0.82, indicating that the model correctly identifies 82% of all 'High' instances in the dataset. Similarly, the 'Low' and 'Medium' classes show reasonable precision and recall scores, demonstrating the model's ability to differentiate between these classes.

Overall Model Performance: The weighted average F1-score of 0.63 indicates a balanced performance across all classes, taking into account both precision and recall. This suggests that the model strikes a good balance between correctly identifying each class ('High', 'Low', 'Medium') and minimizing false positives and false negatives.

# Ranking

In [165]:
# Sort by 'Total Pack' in descending order
customer_sorted = customer.sort_values(by='Total Pack', ascending=False)

# Assign ranks
customer_sorted['Rank'] = range(1, len(customer_sorted) + 1)

# Display the ranked data
print(customer_sorted[['Customer Code', 'Total Pack', 'Rank']])

       Customer Code  Total Pack   Rank
0            2111797   3933750.0      1
1            2111797   3200000.0      2
2            2111797   1675000.0      3
3            2111797   1600000.0      4
4            2111797   1600000.0      5
...              ...         ...    ...
91089        2123054    -37500.0  91105
91105        2117167    -39900.0  91106
91099        2123054    -71875.0  91107
91108        2124345   -120000.0  91108
91106        2119186   -250000.0  91109

[91109 rows x 3 columns]


In [167]:
customer.columns

Index(['Billing Date', 'Customer Code', 'Distributor', 'Material Code',
       'Product Type', 'Pack Type', 'Pack Size', 'YTD Vol (MT)',
       'Sales_Zone_EAST', 'Pack Classification_BP', 'Pack Classification_CP',
       'Pack Size (g)', 'Total Pack', 'Customer Code Encoded',
       'Total Pack Scaled', 'Total Pack Class'],
      dtype='object')

## Filters

In [168]:
import pandas as pd
import numpy as np

# Example DataFrame (replace with your actual data loading code)
data = {
    'Billing Date': pd.date_range(start='2023-01-01', periods=100),
    'Customer Code': np.random.choice(['C001', 'C002', 'C003'], 100),
    'Distributor': np.random.choice(['D001', 'D002', 'D003'], 100),
    'Material Code': np.random.choice(['M001', 'M002', 'M003'], 100),
    'Product Type': np.random.choice(['Type1', 'Type2'], 100),
    'Pack Type': np.random.choice(['PackA', 'PackB'], 100),
    'Pack Size': np.random.randint(50, 500, size=100),
    'YTD Vol (MT)': np.random.uniform(100, 1000, size=100),
    'Sales_Zone_EAST': np.random.choice([0, 1], 100),
    'Pack Classification_BP': np.random.choice(['Yes', 'No'], 100),
    'Pack Classification_CP': np.random.choice(['Yes', 'No'], 100),
    'Pack Size (g)': np.random.randint(10, 200, size=100),
    'Total Pack': np.random.randint(1000, 10000, size=100),
    'Customer Code Encoded': np.random.randint(0, 100, size=100),
    'Total Pack Scaled': np.random.uniform(0, 1, size=100),
    'Total Pack Class': np.random.choice(['Low', 'Medium', 'High'], 100)
}

df = pd.DataFrame(data)

# Randomly choose two columns for filtering (example)
column1 = 'Total Pack'
column2 = 'Pack Size'

# Define filtering conditions based on the chosen columns
condition1 = df[column1] > 5000
condition2 = df[column2] < 200

# Apply AND operation to filter the data
filtered_data = df[condition1 & condition2]

# Display filtered data
print(filtered_data.head())


  Billing Date Customer Code Distributor Material Code Product Type Pack Type  \
2   2023-01-03          C001        D001          M002        Type1     PackB   
3   2023-01-04          C002        D002          M003        Type2     PackB   
5   2023-01-06          C003        D003          M001        Type2     PackB   
6   2023-01-07          C001        D001          M002        Type1     PackB   
8   2023-01-09          C001        D001          M001        Type2     PackA   

   Pack Size  YTD Vol (MT)  Sales_Zone_EAST Pack Classification_BP  \
2        111    964.796772                0                     No   
3         67    893.831872                0                    Yes   
5        184    566.638270                0                     No   
6        185    280.003871                0                     No   
8         68    470.409089                1                    Yes   

  Pack Classification_CP  Pack Size (g)  Total Pack  Customer Code Encoded  \
2             